spring/spring6/spring6-common/src/main/java/org/infinispan/spring/common/provider/SpringCache.java

```diff
@@ -259,4 +265,38 @@ static RuntimeException throwValueRetrievalException(Object key, Callable<?> loa
    public long getWriteTimeout() {
       return writeTimeout;
    }
+
+   @Override
+   public CompletableFuture<?> retrieve(Object key) {
+      return encodedToValueWrapper(nativeCache.getAsync(key));
+   }
+
+   @Override
+   public <T> CompletableFuture<T> retrieve(Object key, Supplier<CompletableFuture<T>> valueLoaderAsync) {
+
+      return nativeCache.getAsync(key).thenApply(v1 -> {
```

Comment:
Just to make sure I am on the same page. Is this supposed to be a lazily computed value and the locking is done to prevent multiple invocations from calling to the valueLoader ?

If so we should use something that doesn't require locks.  The best way to do this async is instead to store a CompletableFuture in the extra map instead of locks. Then store the notion of a computation in that instead.

ConcurrentMap<Object, CompletableFuture<V>> memoizationMap;

Reply:
i don't think so. I think the value is coming already. This methods are used by the annotations in the case of Mono and Flux (reactive)
I think what we seek here is the same logic as the get with the loader, avoiding caching the wrong value unless it's not present

Reply:
Some really dirty pseudo code would be something like this:

```
getAsync from cache
if not null return value
else then compose
  insert new CompletableFuture into memoization map
  getAsync again just in case of async completion
  if not null set value in completable future
  else then compose
    perform computation setting the value/exception to the CompletableFuture
  return completable future
```